app/src/main/java/com/example/android/sunshine/app/ForecastFragment.java

```diff
@@ -158,16 +166,83 @@ public void onItemClick(AdapterView<?> adapterView, View view, int position, lon
         if (savedInstanceState != null && savedInstanceState.containsKey(SELECTED_KEY)) {
             // The listview probably hasn't even been populated yet.  Actually perform the
             // swapout in onLoadFinished.
-            mPosition = savedInstanceState.getInt(SELECTED_KEY);
+            position = savedInstanceState.getInt(SELECTED_KEY);
+            forecastAdapter.setSelection(position);
         }
 
-        mForecastAdapter.setUseTodayLayout(mUseTodayLayout);
+        forecastAdapter.setUseTodayLayout(useTodayLayout);
 
         return rootView;
     }
 
     @Override
+    public void onDestroyView() {
+        recyclerView.clearOnScrollListeners();
+        super.onDestroyView();
+    }
+
+    private ForecastAdapter.OnClickHandler onClickHandler = new ForecastAdapter.OnClickHandler() {
+        @Override
+        public void onClick(long date, ForecastAdapter.ViewHolder holder) {
+            String locationSetting = Utility.getPreferredLocation(getActivity());
+            ((Callback) getActivity()).onItemSelected(
+                    WeatherContract.WeatherEntry.buildWeatherLocationWithDate(locationSetting, date),
+                    holder
+            );
+            ForecastFragment.this.position = holder.getAdapterPosition();
+        }
+    };
+
+    private void setUpParallax(final View parallaxView) {
+        if (parallaxView == null) {
+            return;
+        }
+        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
+            @Override
+            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
+                super.onScrolled(recyclerView, dx, dy);
+                int max = parallaxView.getHeight();
+                if (dy > 0) {
+                    parallaxView.setTranslationY(Math.max(-max, parallaxView.getTranslationY() - dy / 2));
+                } else {
+                    parallaxView.setTranslationY(Math.min(0, parallaxView.getTranslationY() - dy / 2));
+                }
+            }
+        });
```

Comment:
Can we move some of the logic above out of this anonymous class? IMO it would be easier to read if it was like, `recyclerView.addOnScrollListener(new ParallaxTranslator(parallaxView))` with the logic contained with the `ParallaxTranslator` class. Doesn't have to use these exact names of course! 😀